core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java

```diff
@@ -866,6 +866,8 @@ private static void matchFilter(SubQueryRemoveRule rule,
           LogicVisitor.find(RelOptUtil.Logic.TRUE, ImmutableList.of(c), e);
       final Set<CorrelationId>  variablesSet =
           RelOptUtil.getVariablesUsed(e.rel);
+      /* Only consider the correlated variables which originated from this sub-query level*/
+      variablesSet.retainAll(filter.getVariablesSet());
```

Comment:
Agree that this fix is using the existing mechanism to set the variable in Filter operator, I did have questions like what happens when there other operator nodes etc. I want to handle the specific case for this PR(as you mentioned going in the right direction) and handle as we have more use cases (please let me know if it makes sense).

Agree with you that this code might result in a throwing an exception in cases where previous code was generating a plan. However, I encountered a scenario while debugging the JIRA([CALCITE-5716](https://issues.apache.org/jira/browse/CALCITE-5716)) wherein it was not reporting an error but generating a wrong plan. With the fix provided in this PR it is generating a right result (when running a generated plan). Updated the JIRA with the analysis

Reply:
An important case would be the simple one, `Filter` with a correlated condition, and no correlation variable in `Filter`. This case would break I guess.

There is a trick way to avoid this, we only do this when `filter.getVariablesSet()` is not empty, then we can avoid this. What do you guys think?

Reply:
Yeah, it makes sense to add that changes to be on the safe side. I have changed the code. thanks